Minimal TypeScript example.

QuantumInformation opened the pull request:
The beginnings of typescript!

Handy for working with AWS sdk.

QuantumInformation commented:
@pmuens What do you think mate? Maybe it would be clearer as a sibling to the node rest example? I do feel TypeScript has a benefit as node+AWS usage.

pmuens commented:
Nice! Thanks for that @QuantumInformation 🎉 

🤔 I think this should be at the root of the repository. Maybe something like `aws-node-typescript-rest-api-with-dynamodb`. So the naming is `<provider>-<runtime>-<language>-<projectName>`.

What do you think @nikgraf @DavidWells ?

DavidWells commented:
Yeah a typescript example deserves its own top level folder =)

QuantumInformation commented:
Ok, eta weekend)

QuantumInformation commented:
what about aws-node-rest-typescript-api-with-dynamodb so it sits closer together in the project:
![image](https://cloud.githubusercontent.com/assets/216566/23275234/a9af7614-f9fd-11e6-857b-f2c8b907ede2.png)

pmuens commented:
Thanks for the update @QuantumInformation  👍 

IMHO I would still take the name `aws-node-typescript-rest-api-with-dynamodb`. Here's why:

- This way all the typescript project will be near each other
- IMO the REST part belongs to API and not to TypeScript
- IMO The TypeScript part belongs to the runtime so it's <provider>-<runtime>-<optionalLanguage>-<projectName>

Just my two cents 💰

QuantumInformation commented:
OK guys, its been moved..

pmuens commented:
Great! Thanks @QuantumInformation 👍

QuantumInformation commented:
Any plans to merge this?

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/serverless.yml:
```diff
@@ -0,0 +1,48 @@
+          path: todos
+          method: post
+          cors: true
+
```
This space can be removed

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/tsconfig.json:
```diff
@@ -0,0 +1,16 @@
+    "aws-sdk"
+  ]
+}
+
```
This extra space can be removed

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/create.ts:
```diff
@@ -0,0 +1,45 @@
+
+const uuid = require('uuid');
+import {DynamoDB} from 'aws-sdk'
+//const AWS = require('aws-sdk'); // eslint-disable-line import/no-extraneous-dependencies
```
This should be removed

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/create.ts:
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const uuid = require('uuid');
+import {DynamoDB} from 'aws-sdk'
```
`{DynamoDB}` --> `{ DynamoDB }`

QuantumInformation replied:
Should we add an .editorconfig?

QuantumInformation replied:
and lint rule: 
![image](https://user-images.githubusercontent.com/216566/28638298-5ecb0ad4-723c-11e7-9e1c-fa8b404b3d69.png),
mind you, would need to add tslint.

pmuens replied:
> Should we add an .editorconfig?

> and lint rule

Yes, that sounds like a good idea 👍

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/create.ts:
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const uuid = require('uuid');
```
`import uuid from 'uuid'`

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/README.md:
```diff
@@ -0,0 +1,20 @@
+You can then run the compiler by running `tsc` in this directory. It will pull the settings from .tsconfig and extra @types
+from package.json. The output create.js file is what will be uploaded by serverless.
+
+#
```
Is this headline missing? / can it be removed?

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/README.md:
```diff
@@ -0,0 +1,20 @@
+
+TypeScript (ts) offers type safety which is helpful when working with the AWS SDK, which comes with ts definitions (d.ts)
+
+#compiling
```
Should be `# Compiling`

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/README.md:
```diff
@@ -0,0 +1,20 @@
+#Introduction
```
Should be `# Introduction`

pmuens commented on aws-node-typescript-rest-api-with-dynamodb/serverless.yml:
```diff
@@ -0,0 +1,48 @@
+
+provider:
+  name: aws
+  runtime: nodejs4.3
```
Runtime could be updated to `6.10`

QuantumInformation commented:
I'll do this ASAP as I'm giving a big talk on it soon)

pmuens commented:
> I'll do this ASAP as I'm giving a big talk on it soon)

Great! Thank you very much @QuantumInformation 👍

Let us know if you need any help since you might have some time-pressure due to the talk.

BTW. In the meantime Graphcool released a TypeScript plugin which plays nicely with Serverless: https://github.com/graphcool/serverless-plugin-typescript (in case you don't know it already).

pmuens commented:
Thank you very much for updating the PR @QuantumInformation 👍

Just some minor things (the build is currently failing) and the PR should be in a good state to be merged!

Do you want to add the `editorconfig` and `linting` rules as well?

QuantumInformation commented:
Yes I'll add them in.

QuantumInformation commented:
All done..

pmuens commented:
Well, ok. Looks like there's another minor issue.

Could you please update the README so that the build passes 😬

Thanks in advance @QuantumInformation (we're almost there!).

QuantumInformation commented:
I tried serverless deploy with this example, but get:

![image](https://user-images.githubusercontent.com/216566/28747858-c637febe-74a0-11e7-9905-864bc8cd8083.png)

pmuens commented:
Hey @QuantumInformation thanks for updating the PR again 👍

🤔 interesting... Could you update your Serverless version (we've just published a patch release which fixes a problem with the automatic dev dependency exclusion).

Just a quick question: So you're uploading the transpiled function code, right?

QuantumInformation commented:
Looks good to me:

![image](https://user-images.githubusercontent.com/216566/28754375-cdd8d008-753b-11e7-8aad-1d28a798fd42.png)
maybe it shouldn't include the ts file?

QuantumInformation commented:
I'm using serverless 7.8:

![image](https://user-images.githubusercontent.com/216566/28754392-24d3115c-753c-11e7-9223-d29e60e4e7c9.png)

QuantumInformation commented:
I tried deleting the ts file, same error.

darrenhgc commented:
If I'm not mistaken, the serverless.yml expects the create file to be in a todos directory

QuantumInformation commented:
@pmuens hi can I connect via skype or screen hero about this in prep for the talk tomorrow?

QuantumInformation commented:
@darrenhgc it is the case in this PR.

QuantumInformation commented:
Adding `  region: eu-west-2` resulted in no error.

QuantumInformation commented:
I tried to test the endpoint:

`curl -X POST https://6gyss4e9jh.execute-api.eu-west-2.amazonaws.com/dev/todos --data '{ "text": "Learn Serverless" }'`

but get this error

```
nable to import module 'todos/create': Error
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
END RequestId: 5478f9cc-7638-11e7-92b3-139fc9c0a56e
REPORT RequestId: 5478f9cc-7638-11e7-92b3-139fc9c0a56e	Duration: 9.66 ms	Billed Duration: 100 ms 	Memory Size: 1024 MBMax Memory Used: 19 MB

START RequestId: 6fd5385f-7638-11e7-ab5c-fba339c63863 Version: $LATEST
Unable to import module 'todos/create': Error
```

I've included the generated js

QuantumInformation commented:
Hmm, the js doesn't pass lint.

QuantumInformation commented:
@darrenhgc sorry your right! I'm getting muddled with all the dirs.

Next issue is the js that the tsc generates:

```js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const uuid_1 = require("uuid");
const aws_sdk_1 = require("aws-sdk");
const dynamoDb = new aws_sdk_1.DynamoDB.DocumentClient();
module.exports.create = (event, context, callback) => {
    const timestamp = new Date().getTime();
    const data = JSON.parse(event.body);
    if (typeof data.text !== 'string') {
        console.error('Validation Failed');
        callback(new Error('Couldn\'t create the todo item.'));
        return;
    }
    const params = {
        TableName: process.env.DYNAMODB_TABLE,
        Item: {
            id: uuid_1.default.v1(),
```

```
2017-07-31 23:08:51.021 (+01:00)	d51ad49d-763c-11e7-846c-ef19c264eee1	TypeError: Cannot read property 'v1' of undefined
    at module.exports.create (/var/task/todos/create.js:17:31)```

QuantumInformation commented:
So the last thing I tried is

`curl -X POST https://6gyss4e9jh.execute-api.eu-west-2.amazonaws.com/dev/todos --data '{ "text": "Learn Serverless" }'`

but now I'm getting

```
START RequestId: 090b9d20-7640-11e7-b677-cdbdc5a645f1 Version: $LATEST
2017-07-31 23:31:46.120 (+01:00)	090b9d20-7640-11e7-b677-cdbdc5a645f1	{ ResourceNotFoundException: Requested resource not found
    at Request.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/json.js:48:27)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:105:20)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:77:10)
    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:678:14)
    at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)
    at AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)
    at /var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)
    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:680:12)
    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:115:18)
  message: 'Requested resource not found',
  code: 'ResourceNotFoundException',
  time: 2017-07-31T22:31:46.119Z,
  requestId: 'NIPIBGA54LSUNUKITHSQBIG9QNVV4KQNSO5AEMVJF66Q9ASUAAJG',
  statusCode: 400,
  retryable: false,
  retryDelay: 34.56349359159895 }
2017-07-31 23:31:46.121 (+01:00)	090b9d20-7640-11e7-b677-cdbdc5a645f1	{"errorMessage":"Couldn't create the todo item.","errorType":"Error","stackTrace":["Response.dynamoDb.put (/var/task/todos/create.js:29:22)","Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:360:18)","Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:105:20)","Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:77:10)","Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:678:14)","Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)","AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)","/var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10","Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)","Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:680:12)"]}
END RequestId: 090b9d20-7640-11e7-b677-cdbdc5a645f1
REPORT RequestId: 090b9d20-7640-11e7-b677-cdbdc5a645f1	Duration: 97.77 ms	Billed Duration: 100 ms 	Memory Size: 1024 MBMax Memory Used: 38 MB
```

I'm off to sleep, maybe catchup tomorrow.

QuantumInformation commented:
can't see the dynamo table after serverless deploy either

![image](https://user-images.githubusercontent.com/216566/28803159-630f0662-7652-11e7-9b76-b23ac891500d.png)

pmuens commented:
> @pmuens hi can I connect via skype or screen hero about this in prep for the talk tomorrow?

Hey @QuantumInformation I'm sorry , I'm currently on vacation for two weeks and only have gotten Wifi which is quite spotty. 😞

However I looked into the code and here are some observations:

The handler for you're function is pointing at `todos/create.create` so AWS assumes that there's a `todos` directory with a `create.js` file.

Looking at the dir structure in https://github.com/serverless/examples/pull/93#issuecomment-318905362 shows that there's no `todos` directory but the `create.js` file is in the root. So the handler must be update to `create.create`.

Regarding the DynamoDB table: Can you look into the AWS console and see if another table with the same name is already there? The `DeletionPolicy: Retain` config means that the table won't be removed when you delete the service.

I'll follow this thread via E-Mail notifications and chime in again when I have internet access again. Feel free to comment if this resolves your problem!

QuantumInformation commented:
Thx @pmuens, I've checked all my AWS regions and can't find any tables with the same name, I deleted the last one last night.  I also moved the create.js to /todos

pmuens commented:
> Thx @pmuens, I've checked all my AWS regions and can't find any tables with the same name, I deleted the last one last night. I also moved the create.js to /todos

Alright. Can you share the exact error message you get right now when you deploy. Furthermore can you run `Serverless deploy --verbose` so that we can see the CloudFormation output during deployment?

Furthermore it looks like the `uuid` package is not included properly since the call to the `v1` function fails.

Thanks in advance!

QuantumInformation commented:
I resolved it! I change the name to `service: serverless-typescript-rest-api-with-dynamodb` done a sls remove on the old one and pushed data!
thx for helping!

![image](https://user-images.githubusercontent.com/216566/28814566-052b6790-7696-11e7-8a6f-b648745e028e.png)

pmuens commented:
> I resolved it!

@QuantumInformation great! Glad to hear that you were able to resolve it! Merging... :shipit:
